### PR TITLE
POC: FastAPI

### PR DESCRIPTION
**Pull request by iursevla**
See: https://github.com/blakeblackshear/frigate/pull/13324  and https://github.com/blakeblackshear/frigate/discussions/11915 for more information

# FastAPI POC

This is a POC of using [FastAPI](https://fastapi.tiangolo.com/) to replace Flask in Frigate HTTP API. It is using the defaults provided by FastAPI ([uvicorn](https://www.uvicorn.org/) and [starlette](https://www.starlette.io/)).  This was discussed [here](https://github.com/blakeblackshear/frigate/discussions/11915) and [previous PR](https://github.com/blakeblackshear/frigate/pull/13324) was made.

I'm proposing that we start moving to FastAPI by mounting Flask as a sub-application [sub-application](https://fastapi.tiangolo.com/advanced/sub-applications/#mount-the-sub-application) + [2](https://fastapi.tiangolo.com/advanced/wsgi/?h=flask#including-wsgi-flask-django-others). This means that the entire application still works as is and we can then start moving endpoint by endpoint to FastAPI until we get to a point where all endpoints are converted. At that point, we can remove Flask altogether.

I've started the conversion of 2-3 endpoints to showcase how it works:

* Preview endpoints:
	* `/preview/{camera_name}/start/{start_ts}/end/{end_ts}`
    * `/preview/{camera_name}/start/{start_ts}/end/{end_ts}/frames`
* Media endpoints:
	* `/{camera_name}/latest.{extension}` -> See warning below
* App endpoints:
	 * `/` -> Health
	 * `/config/schema.json` 
* Log endpoints:
	 * `/logs/{service}` 	

<details>
<summary>Here it is working with FastAPI </summary>

https://github.com/user-attachments/assets/da648364-1388-4587-9902-b0cc2fa42197

</details>

---

### HTTP API Documentation

<details>
<summary>Details</summary>

Without any extra effort, the swagger documentation is available at `http://localhost:5001/docs` or `http://localhost:5001/redoc` [See documentation](https://fastapi.tiangolo.com/tutorial/metadata/#docs-urls)

image

image

https://github.com/user-attachments/assets/de77e51c-ed11-454a-b0e0-8ed3d12e2e22

Screenshot 2024-08-24 at 16 05 59

The JSON/YAML schema can then be used to feed the [HTTP API page](https://docs.frigate.video/integrations/api) without any human interaction.

From this documentation we can quickly create a client to be used by the front end. This can be done automatically with [openapi-generator-cli](https://github.com/OpenAPITools/openapi-generator-cli). I've used this successfully multiple times at work and also for personal projects, e.g., [Frigate HTTP API](https://github.com/intruder-detection/frigate-http-api-typescript) + [2](https://intruder-detection.github.io/frigate-http-api-typescript/).

</details>

---

### Problems faced

> [!CAUTION]
>  I've faced a problem with the existing endpoints with an extension suffix: `.(jpg|jpeg|png|webp|gif)`.As a result ⚠️ [This lines](https://github.com/blakeblackshear/frigate/pull/13615/files#diff-bff60ae2e26f21d7ae405c36fd8737743e953e4f9c4f2824dfbfe5a8fc30087fR225-R234) in `nginx.conf` was commented out. I've left a comment explaining the errors. If someone wants/can help me figure out why it is failing, then use my branch to debug the endpoints that contain an extension.
I've already ported all [media-related](https://github.com/iursevla/frigate-fork-old/blob/poc-fastapi-continued/frigate/api/media.py) endpoints [in this branch](https://github.com/iursevla/frigate-fork-old/tree/poc-fastapi-continued) and also as a [PR ready to target this current PR](https://github.com/iursevla/frigate-fork/pull/4). If somebody wants to test the problem with the endpoint terminated with those extensions use the mentioned branch.

### CODE REVIEW

File: frigate/app.py [hunk 1]
```diff
@@ -397,6 +400,15 @@ def init_web_server(self) -> None:
...
         )
 
+        self.fastapi_app = create_fastapi_app(
+            self.config,
+            self.detected_frames_processor,
+            self.storage_maintainer,
+            self.onvif_controller,
+            self.plus_api,
+            self.stats_emitter,
+        )
```
Comment (iursevla) [inline frigate/app.py:410]: @NickM-27 @hawkeye217 @blakeblackshear When you have some time please take a look. Please let me know what you think about this PR.

Comment (NickM-27) [reply frigate/app.py:410]: will do, we are currently really busy with ffmpeg 7

Comment (iursevla) [reply frigate/app.py:410]: Got it, @NickM-27, totally understandable. I hope you guys get it working 💯 . 

I'll continue working on other endpoints in my free time. I already covered all **media** endpoints, and today I will convert **notification** and **preview** (at least).

So, in short:

* POC -> this PR (let's call it B1)
* media endpoints -> [This branch](https://github.com/iursevla/frigate-fork/tree/fastapi-poc-media-endpoints) (B2)
* notification and preview  -> on it (B3)

Basically dev <- B1 <- B2 <- B3 <- B.....

Comment (iursevla) [reply frigate/app.py:401]: I got to excited and converted it all on the third branch :D 

Here's the PR: https://github.com/blakeblackshear/frigate/pull/13871

**Comment by iursevla** [comment]
Will update the tests and add docusaurus for the HTTP API. I'm facing some problems with openapi but will get there: https://github.com/PaloAltoNetworks/docusaurus-openapi-docs/issues/980
